hail/python/hail/experimental/vcf_combiner.py

```diff
@@ -1,73 +1,88 @@
 """A work in progress pipeline to combine (g)VCFs into an alternate format"""
 
 import hail as hl
-from hail.matrixtable import MatrixTable
+from hail import MatrixTable, Table
 from hail.expr import StructExpression
 from hail.expr.expressions import expr_call, expr_array, expr_int32
-from hail.ir import TableKeyBy
+from hail.ir import Apply, TableKeyBy, TableMapRows, TopLevelReference
 from hail.typecheck import typecheck
 
+_transform_rows_function_map = {}
+_merge_function_map = {}
```

Comment:
What's the point of having these maps? Can you track of the transformations on different types of rows without this global state? (If I'm interpreting their use correctly)

Reply:
As you surmise, it's to track the input types across function calls. I don't know of any way to track types we have already seen before across top level function calls without some global state. Every call to `define_function` produces a new item in the IR Function Registry. It's essential to keep this down otherwise we lose any of the benefits of defining the functions in the first place.

Does this make sense? I could probably get rid of the global state, it would be a larger change. Would you like me to?

Reply:
I guess it seems more intuitive to me to call some `transform` method that would map `transform_one` over a list of matrix tables and you can cache functions there. Would you be using `transform_one` in multiple places in such a way that having the global state would be advantageous?

Reply:
Yes. gVCFs are merged 100+ files at a time. From what I think you are suggesting, we would only cache on the level of those 100-200 files, which would generate a few thousand identical functions per pipeline.

One solution would be to re-architect the joint caller to be an object, like so. I can do that, but then I would roll all these 
```python3
class GVCFCombiner(object):
    def __init__(self):
        self._transform_fun_map = {}
        self._merge_fun_map = {}

    def transform_one(self, mt) -> Table:
        ...

    def combine_gvcfs(self, mts) -> MatrixTable:
        ...
```
These would be the only two functions that need to be rolled into the new object, but I could do the rest as static methods. Now, this just pushes the issue downstream, because I can't think of why anyone would want more than one joint caller object, and so, global state here vs global state somewhere else.

I'm not sure if the API would be simpler/more explicit this way. Especially since a lot of the explicitness can be solved with proper naming of an import.
```python
import hail.experimental.vcf_combiner as joint_caller
```

Reply:
Hm, ya that does just push the global state downstream, I'm fine with keeping it as it is. Great explanation, thanks!